src/merkle_tree/constraints.rs

```diff
@@ -1,40 +1,87 @@
 use crate::crh::TwoToOneCRHGadget;
-use crate::merkle_tree::Config;
+use crate::merkle_tree::{Config, IdentityDigestConverter};
 use crate::{CRHGadget, Path};
 use ark_ff::Field;
 use ark_r1cs_std::alloc::AllocVar;
 use ark_r1cs_std::boolean::Boolean;
-use ark_r1cs_std::eq::EqGadget;
 #[allow(unused)]
 use ark_r1cs_std::prelude::*;
 use ark_r1cs_std::ToBytesGadget;
 use ark_relations::r1cs::{Namespace, SynthesisError};
 use ark_std::borrow::Borrow;
+use ark_std::fmt::Debug;
 use ark_std::vec::Vec;
+
+pub trait DigestVarConverter<From, To> {
+    fn convert(from: From) -> Result<To, SynthesisError>;
+}
+
+impl<T> DigestVarConverter<T, T> for IdentityDigestConverter<T> {
+    fn convert(from: T) -> Result<T, SynthesisError> {
+        Ok(from)
+    }
+}
```

Comment:
I guess this wrapper is not needed anymore, or?

Reply:
In fact I’m still thinking whether it’s possible to connect leaf hash and inner hash without converting them to byte. This is no a wrapper: instead we let the user of merkle decide how to connect those hashes by specifying a way to convert leaf hash output to inner hash output. 

If user prefer a snark friendly hash, they can provide a function to convert the leaf digest to field elements and convert field elements to inner digest (or anything more direct). Does it looks good?

Reply:
mark as resolved